### PR TITLE
start attempt to rustup to https://github.com/rust-lang/rust/pull/60462

### PR DESCRIPTION
**Pull request by matthiaskrgr**
I am probably too tired to continue working on this right now, in case anyone else wants to pick this up, I put my current progress out there.

This is an attempt to rustup to https://github.com/rust-lang/rust/pull/60462

changelog: none

### CODE REVIEW

File: clippy_lints/src/consts.rs [hunk 1]
```diff
@@ -247,7 +247,7 @@ impl<'c, 'cc> ConstEvalLateContext<'c, 'cc> {
                 if_chain! {
                     if args.is_empty();
                     if let ExprKind::Path(qpath) = &callee.node;
-                    let def = self.tables.qpath_def(qpath, callee.hir_id);
+                    let def = self.tables.qpath_res(qpath, callee.hir_id);
```
Comment (eddyb) [inline clippy_lints/src/consts.rs]: Also rename variables from `def` to `res`.

File: clippy_lints/src/consts.rs [hunk 2]
```diff
@@ -322,9 +322,9 @@ impl<'c, 'cc> ConstEvalLateContext<'c, 'cc> {
     fn fetch_path(&mut self, qpath: &QPath, id: HirId) -> Option<Constant> {
         use rustc::mir::interpret::GlobalId;
 
-        let def = self.tables.qpath_def(qpath, id);
+        let def = self.tables.qpath_res(qpath, id);
```
Comment (eddyb) [inline clippy_lints/src/consts.rs]: Same here.

File: clippy_lints/src/loops.rs [hunk 3]
```diff
@@ -2380,13 +2380,13 @@ impl<'a, 'tcx> VarCollectorVisitor<'a, 'tcx> {
         if_chain! {
             if let ExprKind::Path(ref qpath) = ex.node;
             if let QPath::Resolved(None, _) = *qpath;
-            let def = self.cx.tables.qpath_def(qpath, ex.hir_id);
+            let def = self.cx.tables.qpath_res(qpath, ex.hir_id);
```
Comment (eddyb) [inline clippy_lints/src/loops.rs]: Same here.

File: clippy_lints/src/misc.rs [hunk 4]
```diff
@@ -600,9 +601,9 @@ fn in_attributes_expansion(expr: &Expr) -> bool {
 }
 
 /// Tests whether `def` is a variable defined outside a macro.
-fn non_macro_local(cx: &LateContext<'_, '_>, def: &def::Def) -> bool {
+fn non_macro_local(cx: &LateContext<'_, '_>, def: &def::DefKind) -> bool {
```
Comment (eddyb) [inline clippy_lints/src/misc.rs]: This probably wants to be `res: def::Res` (also update the comment).
Oh, and, small nit: there's no need to pass any of these values by reference.

File: clippy_lints/src/no_effect.rs [hunk 5]
```diff
@@ -70,9 +70,9 @@ fn has_no_effect(cx: &LateContext<'_, '_>, expr: &Expr) -> bool {
         },
         ExprKind::Call(ref callee, ref args) => {
             if let ExprKind::Path(ref qpath) = callee.node {
-                let def = cx.tables.qpath_def(qpath, callee.hir_id);
+                let def = cx.tables.qpath_res(qpath, callee.hir_id);
```
Comment (eddyb) [inline clippy_lints/src/no_effect.rs]: Same here.

File: clippy_lints/src/no_effect.rs [hunk 6]
```diff
@@ -153,9 +153,9 @@ fn reduce_expression<'a>(cx: &LateContext<'_, '_>, expr: &'a Expr) -> Option<Vec
...
```
Comment (eddyb) [inline clippy_lints/src/no_effect.rs]: Hmm, I expected less of these, maybe I should just wait.

File: clippy_lints/src/utils/mod.rs [hunk 7]
```diff
@@ -213,7 +213,7 @@ pub fn match_path_ast(path: &ast::Path, segments: &[&str]) -> bool {
 }
 
 /// Gets the definition associated to a path.
-pub fn path_to_def(cx: &LateContext<'_, '_>, path: &[&str]) -> Option<def::Def> {
+pub fn path_to_def(cx: &LateContext<'_, '_>, path: &[&str]) -> Option<(def::Res)> {
```
Comment (eddyb) [inline clippy_lints/src/utils/mod.rs]: `path_to_res`

**Comment by Manishearth** [comment]
Moved to https://github.com/rust-lang/rust-clippy/pull/4061 . You should have push access.

Also addressed eddyb's nits.

**Pull request closed without merging**